platform/ios/src/MGLCompactCalloutView.m

```diff
@@ -1,31 +1,127 @@
+
+@property (strong, nonatomic) UIButton *mainBody;
+
+@end
 
 @implementation MGLCompactCalloutView
+
++ (instancetype)calloutView
 {
-    id <MGLAnnotation> _representedObject;
+    return [[self alloc] init];
 }
 
-@synthesize representedObject = _representedObject;
-
-+ (instancetype)platformCalloutView
+- (instancetype)initWithFrame:(CGRect)frame
 {
-    return [[self alloc] init];
+    self = [super initWithFrame:frame];
+    if (self)
+    {
+        self.backgroundColor = [UIColor clearColor];
+
+        // Create and add a subview to hold the callout’s text
+        UIButton *mainBody = [UIButton buttonWithType:UIButtonTypeSystem];
+        mainBody.backgroundColor = [self backgroundColorForCallout];
+        mainBody.contentEdgeInsets = UIEdgeInsetsMake(10.0, 10.0, 10.0, 10.0);
+        mainBody.layer.cornerRadius = 4.0;
+        self.mainBody = mainBody;
+
+        [self addSubview:self.mainBody];
+    }
+
+    return self;
 }
 
-- (void)setRepresentedObject:(id <MGLAnnotation>)representedObject
+#pragma mark - MGLCalloutView API
+
+- (void)presentCalloutFromRect:(CGRect)rect inView:(UIView *)view constrainedToView:(UIView *)constrainedView animated:(BOOL)animated
 {
-    _representedObject = representedObject;
-    
-    if ([representedObject respondsToSelector:@selector(title)])
+    // Do not show a callout if there is no title set
+    if (!self.title)
     {
-        self.title = representedObject.title;
+        return;
     }
-    if ([representedObject respondsToSelector:@selector(subtitle)])
+
+    [view addSubview:self];
+
+    // Prepare title label
+    [self.mainBody setTitle:self.title forState:UIControlStateNormal];
+    [self.mainBody sizeToFit];
+
+    // Disable tapping and highlighting
+    self.mainBody.userInteractionEnabled = NO;
+
+    // Prepare our frame, adding extra space at the bottom for the tip
+    CGFloat frameWidth = self.mainBody.bounds.size.width;
+    CGFloat frameHeight = self.mainBody.bounds.size.height + tipHeight;
+    CGFloat frameOriginX = rect.origin.x + (rect.size.width/2.0) - (frameWidth/2.0);
+    CGFloat frameOriginY = rect.origin.y - frameHeight;
+    self.frame = CGRectMake(frameOriginX, frameOriginY,
+                            frameWidth, frameHeight);
+
+    if (animated)
     {
-        self.subtitle = representedObject.subtitle;
+        self.alpha = 0.0;
+
+        [UIView animateWithDuration:0.2 animations:^{
+            self.alpha = 1.0;
+        }];
     }
 }
 
+- (void)dismissCalloutAnimated:(BOOL)animated
+{
+    if (self.superview)
+    {
+        if (animated)
+        {
+            [UIView animateWithDuration:0.2 animations:^{
+                self.alpha = 0.0;
+            } completion:^(BOOL finished) {
+                [self removeFromSuperview];
+            }];
+        }
+        else
+        {
+            [self removeFromSuperview];
+        }
+    }
+}
+
+
+#pragma mark - Custom view styling
+
+- (UIColor *)backgroundColorForCallout
```

Comment:
Since we’re overriding `-drawRect:`, we can use the existing `backgroundColor` property to indicate the visible portion’s background color.